Cursor aware Component Browser

vitvakatu opened the pull request:
### Pull Request Description

Closes #5220

This PR implements the cursor-aware behavior of the CB, as described in [pivotal issue](https://www.pivotaltracker.com/n/projects/2539304/stories/183521918)


https://user-images.githubusercontent.com/6566674/221807206-39f93cb4-8253-421d-a33a-33ac0aa56e54.mp4


https://user-images.githubusercontent.com/6566674/223124947-259153ca-e656-4349-87b5-47c06fd21af2.mp4


### Important Notes

- The `intended_method` of the node's metadata is marked as deprecated, and all usages are removed.
- It seems *all usages of Scala parser are removed from IDE*. We no longer use it to parse documentation for snippets.

This is how the snippets docs look now:

<img width="410" alt="Screenshot 2023-02-28 at 13 18 11" src="https://user-images.githubusercontent.com/6566674/221808028-d69c54e4-2842-4f1c-aa16-781d3f7765a1.png">


### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

farmaazon committed:
Input parsing and tests.

Fixes after rebase

Add some docs

vitvakatu committed:
fix after merge

vitvakatu committed:
Merge branch 'develop' into wip/farmaazon/cursor-aware-cb-183521918

# Conflicts:
#	app/gui/src/controller/searcher.rs

vitvakatu committed:
wip

vitvakatu committed:
wip

vitvakatu committed:
wip

vitvakatu committed:
wip

vitvakatu committed:
wip

vitvakatu committed:
linting

vitvakatu committed:
Merge branch 'develop' into wip/farmaazon/cursor-aware-cb-183521918

# Conflicts:
#	app/gui/src/controller/searcher.rs
#	app/gui/src/presenter/searcher.rs

vitvakatu committed:
self-review & post-merge fixes

vitvakatu commented on app/gui/src/controller/searcher.rs:
```diff
@@ -628,7 +439,8 @@ impl Searcher {
         let module_repr: Rope = module_ast.repr().into();
-        let position = module_repr.offset_to_location_snapped(def_span.end);
+        // TODO: is it correct?
+        let position = module_repr.offset_to_location_snapped(def_span.end - text::ByteDiff(1));
```
Subtracting one byte from position helps to find local variables declared inside `main` function. Apparently the engine thinks that locals are out of scope otherwise and doesn't send them in `completion` response. This issue should be properly investigated separately.

farmaazon replied:
Let's make a GH issue for that and put its number here.

vitvakatu replied:
I fixed the issue completely by searching position before creating a new node.

vitvakatu commented on app/gui/src/controller/searcher.rs:
```diff
@@ -725,24 +537,24 @@ impl Searcher {
+        let old_context = old_input.context().map(|ctx| ctx.into_ast().repr());
+
+        self.invalidate_picked_suggestions();
+        let context_changed = old_context != new_context;
```
I'm not sure if we're doing the correct thing here. Previous code compared expressions, now we compare access chains, and they usually `None`. So we are reloading the list far less frequently now.

farmaazon replied:
This is correct. Before we were filtering the components by expected argument type (and it depended on the left side of the expression), but now the filtering depends only on accessor chain (i.e. the context).

vitvakatu commented on app/gui/src/controller/searcher.rs:
```diff
@@ -2209,64 +1711,57 @@ pub mod test {
+        assert!(are_same(&arg1.entry, &test_var_1));
+        assert!(are_same(&arg2.entry, &test_var_1));
+
+        // TODO
```
I don't understand what's going on in this part of the test, I will fix it or remove it.

vitvakatu replied:
Seems it is not compatible with the new behavior, removed this piece

vitvakatu commented on app/gui/src/controller/searcher/component/hardcoded.rs:
```diff
@@ -106,10 +91,8 @@ impl Snippet {
-        let doc_string = documentation.to_string();
-        let documentation_html = doc_parser.generate_html_doc_pure(doc_string);
-        self.documentation_html = Some(documentation_html.unwrap());
+        // TODO: we need the Scala parser in searcher tests only for HTML documentation.
```
This doesn't look like the correct change, but removing the doc parser usage allows to turn all searcher tests into native ones. This is a huge advantage while developing, though I probably should revert this afterwards and make searcher tests `wasm_test` again.

vitvakatu replied:
At the same time, the documentation for  snippets is extremely primitive and doc parser doesn't really bring much.

farmaazon replied:
We can stay with this change. Before the searcher, tests were usually wasm_bindgen anyway, because we used the Scala parser. But I guess it's no longer the case?

vitvakatu replied:
Yep, now the tests use Rust parser in general.

vitvakatu replied:
I decided to remove the scala parser from here, it does not bring us much.

vitvakatu commented on app/gui/src/controller/searcher/input.rs:
```diff
@@ -0,0 +1,401 @@
+        let range_start = raw_range.start;
+        let range_end = raw_range.end;
+        Self::ensure_on_char_boundary(&new_input, range_start)?;
+        Self::ensure_on_char_boundary(&new_input, range_end)?;
```
I saw rare panics when one of the indices was not on char boundary inside the input, which was weird. This code helps to mitigate the issue if it ever happens again.

farmaazon replied:
I'd "inline" the `range_start` and `range_end` variables. They won't make the lines too long.

vitvakatu commented on app/gui/src/model/module.rs:
```diff
@@ -411,6 +411,7 @@ pub struct NodeMetadata {
     /// The methods may be defined for different types, so the name alone don't specify them.
     #[serde(default, deserialize_with = "enso_prelude::deserialize_or_default")]
+    // TODO: It is not used anywhere. Deprecate?
     pub intended_method: Option<MethodId>,
```
Now all usages of the `intended_method` are removed. We are no longer using it to communicate with the engine, so this field should be probably removed or marked as deprecated.

vitvakatu replied:
Marked it as deprecated and removed all usages from the code.

vitvakatu committed:
wip

vitvakatu committed:
Fix the issue with missing local scope vars

vitvakatu committed:
Minor fixes and refactorings

farmaazon commented on app/gui/src/controller/searcher/input.rs:
```diff
@@ -0,0 +1,401 @@
+impl Input {
+    /// Create the structure from the code already parsed to AST.
+    pub fn new(ast: ast::BlockLine<Ast>, cursor_position: text::Byte) -> Self {
+        let cursor_position_before =
+            (cursor_position > text::Byte(0)).then(|| cursor_position - text::ByteDiff(1));
+        let ast_stack_on_left =
+            cursor_position_before.map(|cp| AstAtPositionFinder::find(&ast.elem, cp));
+        let edited_ast_on_left = ast_stack_on_left.map_or_default(EditedAst::new);
+        let edited_ast = if edited_ast_on_left.edited_name.is_none() {
+            let ast_stack_on_right = AstAtPositionFinder::find(&ast.elem, cursor_position);
+            let edited_ast_on_right = EditedAst::new(ast_stack_on_right);
+            if edited_ast_on_right.edited_name.is_some() {
+                edited_ast_on_right
+            } else {
+                edited_ast_on_left
+            }
+        } else {
+            edited_ast_on_left
+        };
+        let ast = InputAst::Line(ast);
+        Self { ast, cursor_position, edited_ast }
```
This code deserves a bit of explanation in the comment.

farmaazon commented on app/gui/src/controller/searcher/input.rs:
```diff
@@ -0,0 +1,401 @@
+    }
+}
+
+#[derive(Clone, Debug, Fail)]
+#[allow(missing_docs)]
+#[fail(display = "Not a char boundary: index {} at '{}'.", index, string)]
+pub struct NotACharBoundary {
+    index:  usize,
+    string: String,
+}
```
Let's put it in `Errors` section somewhere at the top.

vitvakatu replied:
Done

farmaazon commented on app/gui/src/controller/graph.rs:
```diff
@@ -999,14 +1013,8 @@ impl Handle {
-        // If the name is different than intended method than apparently it is not intended anymore
-        // and should be ignored.
-        let matching = if let Some(name) = name { name == db_entry.name } else { true };
-        matching.then(|| db_entry.invocation_info())
+    fn call_info(&self, _id: node::Id, _name: Option<&str>) -> Option<CalledMethodInfo> {
+        None
     }
 }
```
We already have `Empty` context for such cases: https://github.com/enso-org/enso/blob/d8f6f3fd52d87f63d8b067b609162423ed23ebcc/app/gui/language/span-tree/src/generate/context.rs

So, if we are not in some weird generics' machinery, the controller should not implement Context at all, and use `Empty` instead if itself where needed.

vitvakatu replied:
Done, also removed a couple of tests that relied on  arguments information from node metadata.

farmaazon commented on app/gui/src/test.rs:
```diff
@@ -264,7 +264,7 @@ pub mod mock {
             let mut binary_client = binary::MockClient::new();
             // Creating a searcher controller always triggers a query for completion.
-            controller::searcher::test::expect_completion(&mut json_client, &[]);
+            // controller::searcher::test::expect_completion(&mut json_client, &[]);
```
Commented code. Remove if unneeded.

vitvakatu replied:
Fixed it.

farmaazon commented on app/gui/view/graph-editor/src/component/node/input/area.rs:
```diff
@@ -850,6 +852,8 @@ ensogl::define_endpoints! {
         /// Set the node expression.
         set_expression (node::Expression),
 
+        edit_expression (text::Range<Byte>, ImString),
```
Missing docs.

farmaazon committed:
Apply self-review

vitvakatu committed:
Address review comments & remove broken tests

vitvakatu committed:
Merge branch 'develop' into wip/farmaazon/cursor-aware-cb-183521918

kazcw commented on app/gui/src/controller/searcher.rs:
```diff
@@ -971,25 +714,22 @@ impl Searcher {
+        let expression = self.get_expression(self.data.borrow().input.ast());
         self.graph.graph().set_expression(node_id, expression)?;
-        if let Mode::NewNode { .. } = self.mode.as_ref() {
+        let graph = self.graph.graph();
```
Is this bound for convenience, or to capture the value returned by `graph()` before a subsequent operation changes it? If it's for convenience it should be used consistently so it doesn't look like something tricky is happening; if it is intentionally capturing an older value, a comment would be helpful.

kazcw commented on app/gui/src/controller/searcher/input.rs:
```diff
@@ -0,0 +1,418 @@
+
+    /// Return `true` if the `index` is on a char boundary inside `string` and can be used as a
+    /// range boundary. Otherwise, return `false` and report an error.
+    fn ensure_on_char_boundary(string: &str, index: usize) -> FallibleResult<()> {
```
Docs don't match return type.

kazcw commented on app/gui/src/controller/searcher/input.rs:
```diff
@@ -0,0 +1,418 @@
+            let cursor_position_in_name = self.cursor_position - edited.range.start;
+            let name = ast::identifier::name(&edited.ast);
+            name.map_or("", |n| {
+                &n[text::Range::new(text::Byte(0), text::Byte(0) + cursor_position_in_name)]
```
This slice can panic if `edited_ast` and `cursor_position` are inconsistent. That *shouldn't* happen, but the invariant relied on is not trivial--slicing with `get` here would reduce the consequences of potential future bugs.

kazcw reviewed (approved):
Looks good. @farmaazon I think this answers my questions about how the frontend uses the Scala doc parser!

vitvakatu committed:
Apply review comments

farmaazon commented:
# QA report

## Issues

When starting editing node in the middle of identifier, I see CB with all entries first, and the filtering appears only after a moment.

https://user-images.githubusercontent.com/3919101/222146581-a07528ec-7efd-4f0e-8baa-eed9711ff277.mp4

## Bad UX

Besides the above issue, the PR meets the acceptance criteria. Still, when testing, I found additional things worsening my UX:

* When writing literals (text, number), the results are no longer filtered, so we consistently try to add the first selected entry (usually Column.from_vector) to the input. Before, because literals were part of pattern (and usually no entry matched the literal) the component browser was empty. The solution would be to not display suggestions when the cursor is in or next to a literal, or, even better, we could display a single entry, for example `"foo bar"` if user puts `"foo bar`, or `43` if user types `43` - but this is worth a separate task.
* When editing existing static method invocation, accepting the same method adds the type name again, resulting in erroneous code. It may also be seen as error in visualization preview:

![Screenshot from 2023-03-01 13-50-13](https://user-images.githubusercontent.com/3919101/222149494-d00cc118-a927-4834-8e14-b9325c404269.png)

Solution for it is implementing #5356 we may consider keeping this branch unmerged until we implement and merge that task to it.

vitvakatu committed:
Implement dedicated suggestions for number and text literals

vitvakatu commented:
In commit https://github.com/enso-org/enso/pull/5770/commits/e4b606e394e61f854d5c58334aeacd25f412404c I implemented a requested feature: when writing literals we display a single entry, for example "foo bar" if user puts `"foo bar`, or `43` if user types `43`.


https://user-images.githubusercontent.com/6566674/223124947-259153ca-e656-4349-87b5-47c06fd21af2.mp4

farmaazon reviewed (commented):
The code changes for the new "Literal" entry looks good.

vitvakatu committed:
Improve handling of text literals & add suggestiondb-based documentation

wdanilo reviewed (approved):
I don't have other code-related comments than Adam. However, I've got a few questions - each starts with "do we support it / when we will support it":
- not putting space after replacing a variable. E.g. we have `foo| + bar`, where `|` is mouse cursor. After replacing `foo` with `foo2` we want to see `foo2 |+ bar`.
- Supporting string literals with multiple starting and ending quotes.
- Proper support for negative number literals.

vitvakatu commented:
> not putting space after replacing a variable. E.g. we have foo| + bar, where | is mouse cursor. After replacing foo with foo2 we want to see foo2 |+ bar.

Tracked in https://github.com/enso-org/enso/issues/5848

>    Supporting string literals with multiple starting and ending quotes.

Already implemented. `'''`, `"""`, `'`, `"` are supported as delimiters. Showing documentation for `Text` and `Number` types is also implemented.
    
> Proper support for negative number literals.

Tracked in #5826

vitvakatu committed:
Fix list filtering on CB open

vitvakatu committed:
lint

wdanilo commented:
@vitvakatu add `good first issue` label for issues that you think are simple to be implemented and someone from the community could try implementing. The linked issues look like those to me, but I might be wrong.

vitvakatu committed:
Actually fix the issue with filtering

vitvakatu committed:
lint

vitvakatu committed:
Merge branch 'develop' into wip/farmaazon/cursor-aware-cb-183521918

vitvakatu committed:
Merge branch 'develop' into wip/farmaazon/cursor-aware-cb-183521918

wdanilo commented:
@vitvakatu what is the status of the PR? It was accepted 2 weeks ago (!), and the QA was done then. Why is it not merged yet?

vitvakatu commented:
@wdanilo as we agreed with @farmaazon, we will merge #5356 into this PR, and then merge the whole thing. The QA done 2 weeks ago discovered a few issues, and they were addressed during that two weeks.

wdanilo commented:
@vitvakatu gotcha! Sorry, I forgot about it. Please make short notes about agreements in the future so we will not be missing them. Thanks again! :)

vitvakatu committed:
Merge branch 'develop' into wip/farmaazon/cursor-aware-cb-183521918

vitvakatu committed:
Merge branch 'develop' into wip/farmaazon/cursor-aware-cb-183521918

wdanilo commented:
@vitvakatu as we are making changes here, could you please move it to draft stage and request review again after the code is final, please?

vitvakatu committed:
Implement recognition of qualified names in CB input (#5960)

Closes #5356

Implements recognition of qualified names in the input of the CB. Now accepting of the suggestion produces a valid code, and only necessary imports are added.


https://user-images.githubusercontent.com/6566674/225487445-855dfb04-47c7-44c9-bc92-4bb5663d500a.mp4


https://user-images.githubusercontent.com/6566674/225487486-028d1ae0-02e0-4bbc-a623-7d4e620c87d9.mp4

# Important Notes
- Filtering by inserted code no longer works, but instead, we filter out entries with qualified names not matching the edited accessor chain. It may result in worsened behavior when typing expressions with local variables as `this` argument.

vitvakatu commented:
QA was done in #5960 on an identical state of the code.

mergify[bot] committed:
Merge branch 'develop' into wip/farmaazon/cursor-aware-cb-183521918

vitvakatu committed:
Adjust and disable required imports tests.
